### ISSUE TITLE
craft\web\Controller::requireLogin should be ajax-aware (return JSON)

### ISSUE THREAD

**Issue by timkelty** [craftcms/cms#1451]
### Description
Most core controllers are ajax-aware and will return JSON if requested with ajax. (e.g. entries/save-entry). However if the controller requires a login, you'll simply get back the login page in your response.

Ideally we'd get back a JSON response with an error message.
If this sounds reasonable, I'm happy to PR.

### Steps to reproduce

1. post to a non-anonymous controller (entries/save-entry) via ajax
2. weep

### Additional info

- Craft version: 2.x/3.x

**Comment by timkelty** [comment]
👏  thanks!

**Comment by brandonkelly** [comment]
Yeah so it’s not going to actually return JSON, but at least you won’t get a 200 response anymore. (We did have the option to return JSON, but that option was removed in Yii 2 so decided against it for Craft 2/Yii 1.)

Things are a little different in Craft 3/Yii 2 – the behavior will change based on the `Accepts` header rather than whether it’s an Ajax request. So for example if you’re using `jQuery.ajax()`, you will need to do this:

```javascript
$.ajax({
  // ...
  dataType: 'json'
});
```

Which will send the request with an `Accepts: application/json` header, which will tell `yii\web\User::loginRequired()` to return the 403 error, rather than redirect to the Login page.
